Switch backends to ADTypes.jl

gdalle opened the pull request:
**Main idea**

> Replace our backends with the ones from [ADTypes.jl](https://github.com/SciML/ADTypes.jl) to become compatible with the SciML ecosystem.
> ADTypes.jl is used by Lux.jl, Optimization.jl, NonLinearSolve.jl, OrdinaryDiffEq.jl and Turing.jl, among others.

This reduces the number of LOCs in our package, and only moderately complexifies the custom/fallback mechanism.

**Core**

- Add dependency on ADTypes.jl and use their structs to specify backends
- Remove all our backend code, except for the ChainRules one (which ADTypes.jl doesn't have until https://github.com/SciML/ADTypes.jl/pull/21), renamed to `AutoChainRules`
- Implement trait-based dispatch for forward and reverse mode with `autodiff_mode(backend)`, because some backends from ADTypes.jl can do both (mostly `AutoEnzyme`)
- Replace the `custom` type parameter with another trait-based dispatch:
  - `gradient(backend, f, x)` will call the version using custom package functions
  - `gradient(Val(:fallback), backend, f, x)` will call our fallbacks

Note: the `Val`-based traits can be replaced with custom structs if we want

**Tests**

- Make sure that custom and fallback versions are still tested

**Docs**

- Remove the now useless backends page

**Benchmarks**

- Make sure that custom and fallback versions are still benchmarked
- Add pretty printing

adrhill commented on benchmark/Project.toml:
```diff
@@ -1,9 +1,9 @@
 CSV = "336ed68f-0bac-5ca0-87d4-7b16caf5d00b"
 DataFrames = "a93c6f00-e57d-5684-b7b6-d8193f3e46c0"
 DifferentiationInterface = "a0c0ee7d-e4b9-4e03-894e-1c5f64a51d63"
-Diffractor = "9f5e2b26-1114-432f-b630-d3fe2085c51c"
```
This is due to ADTypes not supporting Diffractor?

gdalle replied:
And Diffractor was failing when I tried to call it through ChainRules. I have opened an issue on the repo, in the meantime I guess we can call it with the AbstractDiff interface as is recommended

adrhill commented on src/array_array.jl:
```diff
@@ -13,23 +13,39 @@ Compute the primal value `y = f(x)` and the Jacobian matrix `jac = ∂f(x)` of a
 
-function _value_and_jacobian!(
+function value_and_jacobian!(
+    ::Val{:fallback},
+    ::Val{:forward},
```
Is there a way to somehow combine this with `extras` we plan on adding? Maybe we could have a parametric struct `Extras{Mode,Implem}`.

adrhill replied:
This `Extras` struct could wrap other `extras` arguments. We could have default constructor for every backend that sets the mode and implementation.

Something like
```Julia
function value_and_jacobian!(jac, backend, f, x, extras)
    value_and_jacobian!(jac, backend, f, x, Extras(backend, extras)
end
```

This way we could avoid boiler-plate code in all other methods besides the lowest-level `value_and_X!` functions by just forwarding `extras`.

gdalle replied:
I don't really like this option because to me the extras should be purely backend specific, and set by the user or by our own "prepare_gradient" methods.
On the other hand, the fallback mechanism is mostly for our internal use and benchmarking, I don't think users should even know about it. After all, if we find that something performs better with our fallbacks, the goal would be to contribute it to the AD package instead of leaving it here.

gdalle replied:
Finally, note that your approach seems to add one method for each function, so it's mostly the same quantity of "boilerplate" as mine. The only difference is for the utilities where we have to differentiate between forward and reverse mode, but then again you would need to do that too

gdalle replied:
However I do think if we keep my approach we should replace value types with custom structs in the trait mechanism, to avoid accidental piracy

gdalle commented:
New proposal here:

**Core**

- The trait `Val{:fallback} / Val{:custom}` has become `FallbackImplem() / CustomImplem()` inheriting from `AbstractImplem`
- The trait `Val{:forward} / Val{:reverse}` has become `ForwardMode() / ReverseMode()` inheriting from `AbstractMode`
- Every utility takes `implem::AbstractImplem` as a first argument.
- In the extensions, every utility must only be defined for `implem::CustomImplem`
- Metaprogramming allows every utility to concisely default on `CustomImplem` (this happens in the `custom.jl` file with `@eval`): no boilerplate :sunglasses: 

**Backend-specific**

- Implemented Zygote without ChainRules using just `Zygote.pullback`
- Added a Diffractor extension but it is broken for non-scalar input

gdalle commented:
Yeah that's a good idea, I'll do that
